### PR TITLE
8331655: ClassFile API ClassCastException with verbose output of certain class files

### PR DESCRIPTION
**Pull request by asotona**
Specifically corrupted constant pool of a class file can cause ClassCastException, when the entries are accessed by Class-File API in exact order.

This fix avoids the ClassCastException and throws ConstantPoolException instead.
Test is attached.

Please review.

Thanks,
Adam

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331655](https://bugs.openjdk.org/browse/JDK-8331655): ClassFile API ClassCastException with verbose output of certain class files (**Bug** - P4)

### Reviewers
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19088/head:pull/19088` \
`$ git checkout pull/19088`

Update a local copy of the PR: \
`$ git checkout pull/19088` \
`$ git pull https://git.openjdk.org/jdk.git pull/19088/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19088`

View PR using the GUI difftool: \
`$ git pr show -t 19088`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19088.diff">https://git.openjdk.org/jdk/pull/19088.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19088#issuecomment-2093250515)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java [hunk 1]
```diff
@@ -399,13 +399,16 @@ public AbstractPoolEntry.Utf8EntryImpl utf8EntryByIndex(int index) {
             int offset = cpOffset[index];
             int tag = readU1(offset);
             final int q = offset + 1;
-            if (tag != TAG_UTF8) throw new ConstantPoolException("Not a UTF8 - index: " + index);
-            AbstractPoolEntry.Utf8EntryImpl uinfo
-                    = new AbstractPoolEntry.Utf8EntryImpl(this, index, buffer, q + 2, readU2(q));
-            cp[index] = uinfo;
-            return uinfo;
+            if (tag == TAG_UTF8) {
```
Comment (PaulSandoz) [inline src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java:402]: Can we call into the tag accepting entryByIndex? e.g.,
```
if (entryByIndex(index, TAG_UTF8) instanceof AbstractPoolEntry.Utf8EntryImpl utf8) {
  return ...
}
throw new ...
```
?
```

Comment (asotona) [reply src/java.base/share/classes/jdk/internal/classfile/impl/ClassReaderImpl.java]: Yes, it is a good opportunity to reduce the code a bit.
Thanks!

**Review by PaulSandoz — APPROVED** [review]
Nice!

**Pull request closed without merging**